### PR TITLE
Add support for 32-, 64-, 128-, 256-, 512-, 1024- bit flavours

### LINKED ISSUE #3: Support higher bit variants
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/fnv1a/issues/3)

FNV1a can be done in 32-, 64-, 128-, 256-, 512-, and 1024-bit flavors. Currently only 32-bit is implemented, because of JS integer limitations. Now that we have `BigInt` in JS, we can implement the other flavors.

More info: http://www.isthe.com/chongo/tech/comp/fnv/index.html#FNV-source

PR welcome :)

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $60.00)

- [issuehunt issuehunt](https://issuehunt.io/u/issuehunt) ($60.00)

### Submitted pull Requests
- [#7 Add BigInt support](https://issuehunt.io/r/sindresorhus/fnv1a/pull/7)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/fnv1a/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
---
IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### PR DESCRIPTION
**Pull request by lunaticmonk**
Fixes #3

### CODE REVIEW

**Comment by sindresorhus** [comment]
Read the issue again. It's mentions `BigInt`. This will also require some tests and docs changes.

**Comment by lunaticmonk** [comment]
@sindresorhus I did some changes. Please review. However, `xo` is throwing linting errors like `BigInt is not defined.`, `A function with a name starting with an uppercase letter should only be used as a constructor.`. Is there a way I can ignore them by modifying some configs?

File: index.js [hunk 1]
```diff
@@ -1,16 +1,51 @@
...
+	32: 2166136261,
+	64: 14695981039346656037,
+	128: getBigInt('144066263297769815596495629667062367629'),
+	256: getBigInt('100029257958052580907070968620625704837092796014241193945225284501741471925557'),
+	512: getBigInt('9659303129496669498009435400716310466090418745672637896108374329434462657994582932197716438449813051892206539805784495328239340083876191928701583869517785'),
+	1024: getBigInt('14197795064947621068722070641403218320880622795441933960878474914617582723252296732303717722150864096521202355549365628174669108571814760471015076148029755969804077320157692458563003215304957150157403644460363550505412711285966361610267868082893823963790439336411086884584107735010676915')
+};
+
+function getBigInt(int) {
+	if (typeof (int) === Number) {
```
Comment (sindresorhus) [inline index.js]: This check is not correct.

File: index.js [hunk 2]
```diff
@@ -1,16 +1,51 @@
...
+		return hash >>> 0;
+	}
+
+	/**
+	 * BigInt calculations for offsetBasis greater than 64.
+	 *
+	 */
+	for (let i = 0; i < string.length; i++) {
+		hash ^= getBigInt(string.charCodeAt(i).toString());
+		hash += (hash << getBigInt('1')) + (hash << getBigInt('4')) + (hash << getBigInt('7')) + (hash << getBigInt('8')) + (hash << getBigInt('24'));
```
Comment (sindresorhus) [inline index.js]: Are you sure the algorithm is exactly the same for larger numbers? Please read up on the source material: http://www.isthe.com/chongo/tech/comp/fnv/index.html#FNV-param

**Comment by sindresorhus** [comment]
I think the 64-, 128-, 256-, 512-, 1024 versions should be a separate method: `fnv1a.bigInt()` which should accept an options-object, so `fnv1a.bigInt({bits: 128})`. The method should validate that only the allowed bit versions are supplied.

File: index.js [hunk 3]
```diff
@@ -1,16 +1,51 @@
...
-	let hash = OFFSET_BASIS_32;
+/**
+ * Offset basis reference: http://www.isthe.com/chongo/tech/comp/fnv/index.html#FNV-param
+ */
 
-	for (let i = 0; i < string.length; i++) {
-		hash ^= string.charCodeAt(i);
+const OFFSET_BASIS = {
+	32: 2166136261,
+	64: 14695981039346656037,
```
Comment (sindresorhus) [inline index.js]: JavaScript does not support 64-bit numbers.

**Comment by lunaticmonk** [comment]
@sindresorhus Okay, will do that. Also, don't you think 64 bit can be solved normally rather than pushing it to bigInt?

**Comment by sindresorhus** [comment]
> Also, don't you think 64 bit can be solved normally rather than pushing it to bigInt?

No, and if you think that, I'm questioning whether you're up to the task to resolve this correctly.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Pull request closed without merging**